Author: Sergio Durigan Junior <sergiodj@chainguard.dev>
Description: Fix FTBFS with Python < 3.12
 .
 Upstream fixed a potential deadlock when using GIL at:
 .
 https://github.com/metaopt/optree/commit/2762fea3f98efbbf5274116f6601e4a0c01f171d
 .
 This introduced some usages of py::gil_scoped_{release,acquire}_simple, which
 don't exist on Python < 3.12.  We have to selectively use the right types
 depending on the Python version we're compiling with.
 .
 It doesn't make sense to upstream this patch as they're targeting Python 3.14
 now.

diff --git a/include/optree/pytypes.h b/include/optree/pytypes.h
index 9e2a949..06b3dca 100644
--- a/include/optree/pytypes.h
+++ b/include/optree/pytypes.h
@@ -30,6 +30,11 @@ limitations under the License.
 
 #if PY_VERSION_HEX < 0x030C00F0  // Python 3.12.0
 #    include <structmember.h>    // PyMemberDef
+#    define GIL_SCOPED_RELEASE_TYPE py::gil_scoped_release
+#    define GIL_SCOPED_ACQUIRE_TYPE py::gil_scoped_acquire
+#else
+#    define GIL_SCOPED_RELEASE_TYPE py::gil_scoped_release_simple
+#    define GIL_SCOPED_ACQUIRE_TYPE py::gil_scoped_acquire_simple
 #endif
 
 #include <pybind11/eval.h>  // pybind11::exec
@@ -306,7 +311,7 @@ public:
         std::optional<StoredType> cached_value{};
         {
 #if !defined(Py_GIL_DISABLED)
-            const py::gil_scoped_release_simple gil_release{};
+            const GIL_SCOPED_RELEASE_TYPE gil_release{};
 #endif
             const scoped_read_lock lock{m_mutex};
             const auto it = m_cache.find(cache_key);
@@ -338,7 +343,7 @@ public:
         bool registered = false;
         {
 #if !defined(Py_GIL_DISABLED)
-            const py::gil_scoped_release_simple gil_release{};
+            const GIL_SCOPED_RELEASE_TYPE gil_release{};
 #endif
             const scoped_write_lock lock{m_mutex};
             const auto [it, inserted] = m_cleanup_registered.try_emplace(interpreter_id, false);
@@ -367,7 +372,7 @@ public:
         bool inserted = false;
         if (registered) [[likely]] {
 #if !defined(Py_GIL_DISABLED)
-            const py::gil_scoped_release_simple gil_release{};
+            const GIL_SCOPED_RELEASE_TYPE gil_release{};
 #endif
             const scoped_write_lock lock{m_mutex};
             if (m_cache.size() < m_max_size) [[likely]] {
diff --git a/src/treespec/traversal.cpp b/src/treespec/traversal.cpp
index ef3a282..3b77e9b 100644
--- a/src/treespec/traversal.cpp
+++ b/src/treespec/traversal.cpp
@@ -174,12 +174,12 @@ py::object PyTreeIter::Next() {
     }
 
 #if !defined(Py_GIL_DISABLED)
-    const py::gil_scoped_release_simple gil_release{};
+    const GIL_SCOPED_RELEASE_TYPE gil_release{};
 #endif
     const scoped_lock lock{m_mutex};
     {
 #if !defined(Py_GIL_DISABLED)
-        const py::gil_scoped_acquire_simple gil_acquire{};
+        const GIL_SCOPED_ACQUIRE_TYPE gil_acquire{};
 #endif
         m_running_thread_id.store(ident, std::memory_order_release);
         try {
